File: hdf-converters/converters/hdf-to-ecs/go/schema_validation_test.go
package hdftoecs

import (
	"encoding/json"
	"fmt"
	"strings"
	"testing"

	corpus "github.com/mitre/hdf-libs/hdf-converters/v3/internal/corpus"
)

// NO DOCUMENT SCHEMA EXISTS, confirmed independently 2026-09-11 rather than
// taken from the audit: elastic/ecs publishes its generated artifacts as
// ecs_flat.yml and ecs_nested.yml (field definitions), Elasticsearch composable
// and legacy index templates (mappings), fields.csv, and Beats fields.ecs.yml.
// None is a JSON Schema. ECS is a field dictionary whose conformance is enforced
// by Elasticsearch mappings at index time, so there is no document-level schema
// to validate an exported event against.
//
// What remains checkable is that every emitted line is a JSON object, which is
// what the NDJSON contract requires. A no-op validator would make every
// MustConvert contract pass vacuously.
type ecsNDJSONValidator struct{}

func (ecsNDJSONValidator) Validate(doc []byte) error {
	if len(doc) == 0 {
		return nil
	}
	for i, line := range strings.Split(strings.TrimSpace(string(doc)), "\n") {
		if line == "" {
			continue
		}
		var obj map[string]interface{}
		if err := json.Unmarshal([]byte(line), &obj); err != nil {
			return fmt.Errorf("line %d is not a JSON object: %w", i+1, err)
		}
		// Unmarshalling the literal null into a map SUCCEEDS and leaves obj nil,
		// so without this a null line passes here while the TypeScript peer
		// rejects it — the two would disagree about the same output.
		if obj == nil {
			return fmt.Errorf("line %d is null, not a JSON object", i+1)
		}
	}
	return nil
}

// TestConvertHDFToECS_AdversarialCorpus holds this converter to the shared
// corpus contracts, with NDJSON well-formedness standing in for a schema.
func TestConvertHDFToECS_AdversarialCorpus(t *testing.T) {
	corpus.RunSchemaCorpus(t, ecsNDJSONValidator{}, corpus.ResultsCorpus(),
		func(in []byte) ([]byte, error) { return ConvertHDFToECS(in, "1.0.0") })
}

File: hdf-converters/converters/hdf-to-ecs/typescript/schema-validation.test.ts
import { describe, it, expect } from 'vitest';
import { resultsCorpus, runSchemaCorpus } from '../../../shared/typescript/schema-corpus.js';
import { convertHdfToEcs } from './converter.js';

// NO PUBLISHED SCHEMA. Confirmed independently 2026-09-11: elastic/ecs publishes ecs_flat.yml and ecs_nested.yml (field definitions), Elasticsearch index templates (mappings), fields.csv and Beats fields.ecs.yml. None is a JSON Schema; ECS conformance is enforced by index mappings, so there is no document schema to validate an event against.
//
// What remains checkable is that every emitted line is a JSON object, which the
// NDJSON contract requires. A no-op validator would make every MustConvert
// contract pass vacuously. Mirrors the Go peer.
const ndjsonObjects = (doc: unknown): string | null => {
  const text = typeof doc === 'string' ? doc : '';
  if (text === '') return null;
  const lines = text.trim().split('\n');
  for (const [i, line] of lines.entries()) {
    if (line === '') continue;
    let parsed: unknown;
    try {
      parsed = JSON.parse(line);
    } catch (e) {
      return `line ${i + 1} is not JSON: ${(e as Error).message}`;
    }
    if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
      return `line ${i + 1} is not a JSON object`;
    }
  }
  return null;
};

describe('ECS adversarial corpus', () => {
  it('satisfies every contract for every case', async () => {
    const cases = resultsCorpus();
    expect(cases.length, 'an empty corpus would pass vacuously').toBeGreaterThan(0);
    await runSchemaCorpus(ndjsonObjects, cases, (input) => convertHdfToEcs(input, '1.0.0'));
  });
});
